### PR TITLE
Fixed #25513 -- Refactored the admin pagination to make it reusable

### PR DESCRIPTION
**Pull request by sasha0**

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by timgraham** [comment]
Could you document the new paginator as noted on the ticket? Also, there are some failing tests.

**Comment by sasha0** [comment]
@timgraham sure, will do.

**Comment by sasha0** [comment]
Also, the thing is that page range in admin pagination starts with 0, not 1 as in `django.core.Paginator`.
If we change this, that would make pagination classes consistent and avoid few not quite necessary increments/decrements, e.g.:
https://github.com/django/django/blob/master/django/contrib/admin/templatetags/admin_list.py#L40
https://github.com/django/django/blob/master/django/contrib/admin/templatetags/admin_list.py#L44-L45
https://github.com/django/django/blob/53ccffdb8c8e47a4d4304df453d8c79a9be295ab/django/contrib/admin/views/main.py#L195
Could we make this change or do we want to keep it as is?

**Comment by timgraham** [comment]
As long as the output is the same, consistency sounds like a good thing.

**Comment by sasha0** [comment]
That's not quite straightforward without modifying existing tests, because reference page range starting from zero:
https://github.com/django/django/blob/master/tests/admin_changelist/tests.py#L743-L748
What do you think?

**Comment by timgraham** [comment]
Modifying tests isn't so much of a problem, but if any user code that overrides things like ModelAdmin methods will be affected, that could be problem. I haven't looked at the details... can you say if this might be the case?

**Comment by sasha0** [comment]
Of course, I'll check.

**Comment by sasha0** [comment]
Hm, found one case - since code which was generating ellipsis pagination was moved from template tag `django.contrib.admin.templatetags.admin_list.pagination` to paginator class `django.core.pagination.EllipsisPaginator`, if we customize paginator class in `ModelAdmin.get_paginator` it could have no logic for generating ellipsis pagination so user would see in admin list output of `CustomPaginator.page_range`.

The thing is that all logic was in template tag, so it was separated from paginator class, thus we could customize paginator and ellipsis paginator without affect to ellipsis pagination.

I've been thinking also about moving ellipsis pagination logic to separate method, however, it would not help with backwards compatibility, since custom paginator classes would not have it.

What do you think?

**Comment by timgraham** [comment]
Maybe we could have a deprecation path so that custom paginators are eventually required to inherit `EllipsisPaginator` or to implement that display method?

**Comment by sasha0** [comment]
Yes, that's quite reasonable - let's stick on separate display method. Thanks!

**Comment by sasha0** [comment]
I've come up with draft version of documentation and moved ellipsis page range generation to separate method. How would you suggest to require presence of method `get_ellipsed_page_range`? 
Does it make sense not to introduce current paginator class, but add this custom method `get_ellipsed_page_range` to `django.core.Paginator`?

File: tests/admin_views/tests.py [hunk 1]
```diff
@@ -2736,7 +2736,7 @@ def test_list_editable_pagination(self):
         self.assertContains(response, 'Unordered object #3')
         self.assertContains(response, 'Unordered object #2')
         self.assertNotContains(response, 'Unordered object #1')
-        response = self.client.get(reverse('admin:admin_views_unorderedobject_changelist') + '?p=1')
+        response = self.client.get(reverse('admin:admin_views_unorderedobject_changelist') + '?p=2')
```
Comment (timgraham) [inline tests/admin_views/tests.py]: This change seems to imply that existing bookmarks will break.

**Comment by timgraham** [comment]
If possible, I think the admin shouldn't require an ellipsed page range. In that case, the method name would be more generic than `get_ellipsed_page_range()` and `Paginator` would have a some suitable generic implementation while `EllipsisPaginator` would override that method. Probably the more functionality that can be moved from the template tag to the `Paginator`, the easier it'll be to customize it.

File: tests/admin_views/tests.py (hunk 1 above)
Comment (sasha0) [inline tests/admin_views/tests.py]: Hm, if you mean, that previous links to admin changelist pages would show another results after proposed change - yes, that's correct.

I raised that concern previously - https://github.com/django/django/pull/5866#issuecomment-168448650.

The reason of the case is simple - pagination page parameter is one less that actual page number, e.g. `reverse('admin:admin_views_unorderedobject_changelist') + '?p=2'` is showing 3rd page.

Comment (timgraham) [inline tests/admin_views/tests.py]: I think backwards compatibility isn't critical here, but a mention in the release notes wouldn't hurt.

**Comment by sasha0** [comment]
> and `Paginator` would have a some suitable generic implementation 

You mean, generic implementation of compact page range? Or abstract method? Please suggest.

**Comment by timgraham** [comment]
Not sure... let's see what the implementation for `EllipsisPaginator` looks like first.

**Comment by auvipy** [comment]
hi @sasha0 are you still into it? or this should be closed for now?

**Comment by carltongibson** [comment]
Hi @sasha0. 

I know this has been here a long time, but it looks quite good. Do you have any appetite for finishing it off?

(One quick thought on the BC issue: **I don't know it it's worth it** but we could put a wrapper class around the pagination that handled the move from 0 to 1 indexed. Such a thing would be easy to test and would keep the adjustments in a single location.)

**Comment by srinivasreddy** [comment]
@sasha0  Hey Can i take this to finishing line?

**Comment by timgraham** [comment]
Continued in https://github.com/django/django/pull/10328.

**Pull request closed without merging**